integration-tests/src/test_loop/tests/fix_min_stake_ratio.rs

```diff
@@ -0,0 +1,138 @@
+        .shard_layout(ShardLayout::get_simple_nightshade_layout_v3())
+        .protocol_version(ProtocolFeature::FixMinStakeRatio.protocol_version() - 1)
+        .epoch_length(epoch_length)
+        .validators_raw(validators, 1, 2)
+        // For genesis, set high minimum stake ratio so that small validator
+        // will be excluded from the validator set.
+        .minimum_stake_ratio(Rational32::new(1, 6_250))
+        // Disable validator rewards.
+        .max_inflation_rate(Rational32::new(0, 1));
+    for account in &accounts {
+        genesis_builder.add_user_account_simple(account.clone(), initial_balance);
+    }
+    let genesis = genesis_builder.build();
+
+    let TestLoopEnv { mut test_loop, datas: node_datas, tempdir } =
+        builder.genesis(genesis).clients(clients).build();
+
+    let client_sender = node_datas[0].client_sender.clone();
+    let client_handle = node_datas[0].client_sender.actor_handle();
+    let initial_validators = get_epoch_all_validators(&test_loop.data.get(&client_handle).client);
+    assert_eq!(initial_validators.len(), 2);
+    assert!(!initial_validators.contains(&small_validator.to_string()));
+
+    // Run chain for couple epochs.
+    // Due to how protocol version voting works, chain must automatically
+    // upgrade to the latest protocol version which includes FixMinStakeRatio
+    // enabled.
+    // In the epoch where it happens, small validator must join the validator
+    // set.
+    let latest_epoch_height = AtomicU64::new(0);
+    let success_condition = |test_loop_data: &mut TestLoopData| -> bool {
+        let client = &test_loop_data.get(&client_handle).client;
+        let validators = get_epoch_all_validators(client);
+        let tip = client.chain.head().unwrap();
+        let epoch_height =
+            client.epoch_manager.get_epoch_height_from_prev_block(&tip.prev_block_hash).unwrap();
+
+        if epoch_height > latest_epoch_height.load(Ordering::Relaxed) {
+            latest_epoch_height.store(epoch_height, Ordering::Relaxed);
+            // At each epoch start, stake with small validator so that it
+            // will be included in the list of proposals.
+            let sender = &accounts[2];
+            let tx = SignedTransaction::stake(
+                epoch_height,
+                sender.clone(),
+                &create_user_test_signer(sender).into(),
+                ONE_NEAR,
+                near_primitives::test_utils::create_test_signer(accounts[2].as_str()).public_key(),
+                tip.prev_block_hash,
+            );
+            let future = client_sender.send_async(ProcessTxRequest {
+                transaction: tx,
+                is_forwarded: false,
+                check_only: false,
+            });
+            drop(future);
+        }
+
+        assert!(epoch_height < 4);
+        return if validators.len() == 3 {
+            assert!(validators.contains(&small_validator.to_string()));
+            true
```

Comment:
nit: there could also be a check `current_epoch.min_stake_ratio == 1/62500` to double check that the validator doesn't get included before stake ratio fix.

Reply:
Nice idea, added.